### PR TITLE
Getting heteropolymers builds working

### PR DESCRIPTION
**Pull request by mrshirts**
* You can now set bond lengths, sigma, and epsilon on a per-monomer basis.  Other energy terms are still sidechain/backbone.  May not be that hard to do it in the future. This is demonstrated in get_heteropolymer_structure()
* bond lengths, sigmas, and epsilons HAVE to be set on a per-monomer basis.  Eventually it would be good to set this on a particle-type basis, and just have the monomers define particles.
* This changes the example script get_initial_structure.py
* This changes what you pass into cgmodel initialization as well.
* No longer automatically creates monomer types in the case of just polymer_length being provided.
* New function "process_monomer_types" to clean up/fill in information.
* More information is contained in the particle list now, which is a dictionary for each particle, and added more utility functions to get information from them.
* Some general cleaning: 
    - ("coarse grained" -> "coarse-grained")
    - setting defaults for cgmodel more cleanly.
    - ran black which changed some previous spacing.
    - cleaned handling of array-ness of "sidechain positions"
    - more f-strings
    - removed functions get_all_particle_masses and get_bond_lengths_from_names that were never used.
    - removed some code that was never reached
   - removed basic_cg_model since it didn't really do anything.

Goes in parallel with https://github.com/shirtsgroup/cg_openmm/pull/44.

### CODE REVIEW

File: examples/generate_random_structures/get_heteropolymer_structure.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    equil_torsion_angles=equil_torsion_angles,
+    torsion_periodicities=torsion_periodicities,
+    include_nonbonded_forces=include_nonbonded_forces,
+    include_bond_forces=include_bond_forces,
+    include_bond_angle_forces=include_bond_angle_forces,
+    include_torsion_forces=include_torsion_forces,
+    constrain_bonds=constrain_bonds,
+    random_positions=random_positions,
+    sequence=sequence,
+    monomer_types=[A, B, C],
```
Comment (cwalker7) [inline examples/generate_random_structures/get_heteropolymer_structure.py:137]: Could we have sequence just be a list of strings ['A','B','C','A',...] instead of list of dictionaries so we don't need to pass all of the info for every monomer? It seems that all the detailed info we need is contained in monomer_types.

File: foldamers/cg_model/cgmodel.py [hunk 2]
```diff
@@ -174,21 +84,15 @@ class CGModel(object):
     def __init__(
```
Comment (cwalker7) [inline foldamers/cg_model/cgmodel.py:84]: Do we need to update the _built_in_regions here to match the changes? Not really sure what this part is for.

File: examples/generate_random_structures/get_heteropolymer_structure.py (hunk 1 above)
Comment (mrshirts) [reply examples/generate_random_structures/get_heteropolymer_structure.py:137]: Good question.  It could be done, but internally, then one would have to translate with a dictionary linking monomer types to monomer characters.  As it is, this translation doesn't need to be done.  It's not really that big of overhead, since I'm pretty sure python is passing by reference anyway (i.e. [A,B,C] I think is a list of pointers to the dictionary members. And it's not that large an amount of data.  But I'm definitely open to switching if it seems like there are other advantages.

Comment (cwalker7) [reply examples/generate_random_structures/get_heteropolymer_structure.py:137]: Ah ok - in that case I think it's fine to leave sequence as is for now.

File: foldamers/cg_model/cgmodel.py [hunk 3]
```diff
@@ -875,45 +750,38 @@ def get_particle_type(self, particle_index, particle_name=None):
...
-            )
-            exit()
+        else:
+            print(f"ERROR: The particle type definition could not be found for {particle_index}")
 
         return particle_type
 
+    def get_particle_monomer(self, particle_index):
+        """
+          Indicates if a particle is a backbone bead or a sidechain bead
```
Comment (cwalker7) [inline foldamers/cg_model/cgmodel.py:768]: This description wasn't changed from get_particle_type

File: foldamers/cg_model/cgmodel.py (hunk 2 above)
Comment (mrshirts) [reply foldamers/cg_model/cgmodel.py:84]: Good question on the _BUILT_IN_REGIONS_.  There IS actually a purpose here, I figured out, but I'm not sure this is used correctly, and may not be useful now.  It's only defined here, and not used in the rest of the system.  Looks like in Yank, one could define built-in regions (see http://getyank.org/latest/_modules/yank/yank.html), but we're not using yank anymore, and in any case those would just be atom selections, not general model properties. So I will modify the PR and remove this.

File: foldamers/cg_model/cgmodel.py (hunk 3 above)
Comment (mrshirts) [reply foldamers/cg_model/cgmodel.py]: OK fixed.

**Comment by cwalker7** [comment]
I ran the heteropolymer builder example with the 4 * ABC sequence and it looks like something is wrong with the connectivity / particle types. There are some side chain beads in the backbone and vice versa. Will have to look further into it.

File: foldamers/cg_model/cgmodel.py [hunk 4]
```diff
@@ -578,10 +416,10 @@ def get_bond_list(self):
...
                         and monomer_type["backbone_length"] - 1
-                        in [monomer_type["sidechain_positions"]]
+                        in monomer_type["sidechain_positions"]
                     ) or (
                         backbone_bead != 0
-                        and backbone_bead - 1 in [monomer_type["sidechain_positions"]]
+                        and (backbone_bead - 1) in monomer_type["sidechain_positions"]
                     ):
                         parent_index = bead_index - monomer_type["sidechain_length"] - 1
                     else:
```
Comment (cwalker7) [inline foldamers/cg_model/cgmodel.py:425]: I think the problem is that it's bonding the backbone beads based on sidechain length of the _current_ monomer type, not the previous which can now have different numbers of sidechain beads.

At any rate we should revisit this logic - it's a bit hard to follow even for homopolymers.

Comment (mrshirts) [reply foldamers/cg_model/cgmodel.py:425]: Good detective work.  Let me try a patch for this.  Let's wait to reorganize the logic until we change it up so we can have arbitrary topology monomers / particle types, since we want to do that relatively soon.

Comment (mrshirts) [reply foldamers/cg_model/cgmodel.py:420]: OK, I think this new commit should fix that.

**Comment by cwalker7** [comment]
Ok, the bond connectivity seems to be fixed. However for the heteropolymer builder there is still some dubious behavior with bonds and angles. I get structures that look like this:

image

This was in vmd with 'autobonds off' to reflect what is written in the connect section.

There is sometimes also an error with collisions from the util.py script, yet the structure is still saved to pdb:
`Error: A model was successfully built, however,
particle collisions were detected.

The shortest nonbonded particle distance is:
0.06201851880131594 nm
The nonbonded particle cutoff distance used for
random structure generation is set to:0.8 A`

**Comment by mrshirts** [comment]
Hi, @cwalker7, for the weird structure, can you post the script used to generate it (or put it in a branch) as well the structure coming out that you visualized?

**Comment by mrshirts** [comment]
> There is sometimes also an error with collisions from the util.py script, yet the structure is still saved to pdb:
> `Error: A model was successfully built, however, particle collisions were detected.
Well, the script seems to be doing what it says, in that a model was built even though particle collisions were detected? I guess the question is if the resulting structure has the correct topology and is a low enough energy to be stable when starting a simulation run.  If so, then it's OK; the script is just supposed to be generating starting points, not good structures.

**Comment by cwalker7** [comment]
Here is the input script, output structure, and output text files. Note that this one didn't produce the collision error.
[heteropolymer_testing_7_6_20.zip](https://github.com/shirtsgroup/foldamers/files/4880380/heteropolymer_testing_7_6_20.zip)

**Comment by mrshirts** [comment]
Topology seems totally fine, which is good.  

But there are three angles that are essentially doubled back on each other, with angle 0.  They are clearly the angles that are spanning 3 residues.   I think (not quite sure yet, need to check) is the reason is that it is not assigning forces to angles that span 3 residues, perhaps because it's not creating those angles.  I'll take a pass at figuring out why.

**Comment by mrshirts** [comment]
A little weirder than that - the particle list was not being initialized correctly, giving some weird results for angles/bonds.  Those particular angles were being assigned backbone-sidechain-backbone angles, which are not a thing, so it used the defaults to create those (it was being printed to the screen, so we all perhaps should have caught that!)  

I fixed the initialization, and the structures look more consistent now, i.e. all the angles now 120 as defined in the input.

**Comment by cwalker7** [comment]
Ah ok, good work! I did find it odd it was displaying the bb-sc-bb angle defaults.

Also I see now why you needed a convention for assigning the monomer-monomer connecting bond types to the first monomer. I think we can discuss further in the issue on heteropolymer builder #25

**Comment by cwalker7** [comment]
I'm good with merging both this one and the corresponding cg_openmm now, but still need to make tests for a lot of the cgmodel functions, I only checked if the actual examples were running correctly.

**Comment by mrshirts** [comment]
> I'm good with merging both this one and the corresponding cg_openmm now, but still need to make tests for a lot of the cgmodel functions,

OK, I'll merge and add issues for these.

**Comment by mrshirts** [comment]
Actually, I guess there already are issues for both cg_openmm and foldamers (#9)
